src/hotspot/share/gc/shared/partialArrayState.hpp

```diff
@@ -91,47 +95,136 @@ class PartialArrayState {
                               size_t index, size_t length,
                               size_t initial_refcount);
 
   // Decrement the state's refcount.  If the new refcount is zero, add the
   // state to the free-list associated with worker_id.  The state must have
   // been allocated by this allocator, but that allocation doesn't need to
   // have been associated with worker_id.
-  void release(uint worker_id, PartialArrayState* state);
+  void release(PartialArrayState* state);
+};
+
+// This class provides memory management for PartialArrayStates.
+//
+// States are allocated using an allocator object. Those allocators in turn
+// may request memory for a state from their associated manager. The manager
+// is responsible for obtaining and releasing memory used for states by the
+// associated allocators.
+//
+// A state may be allocated by one allocator, but end up on the free-list of a
+// different allocator.  This can happen because a task referring to the state
+// may be stolen from the queue where it was initially added.  This is permitted
+// because a state's memory won't be reclaimed until all of the allocators
+// associated with the manager that is ultimately providing the memory have
+// been deleted and the manager is reset.
+//
+// A manager is used in two distinct and non-overlapping phases.
+//
+// - allocating: This is the initial phase.  During this phase, new allocators
+// may be created, and allocators may request memory from the manager.
+//
+// - releasing: When an allocator is destroyed the manager transitions to this
+// phase.  It remains in this phase until all extent allocators associated with
+// this manager have been destroyed.  During this phase, new allocators may not
+// be created, nor may extent allocators request memory from this manager.
+//
+// Once all the associated allocators have been destroyed the releasing phase
+// ends and the manager may be reset or deleted.  Resetting transitions back
+// to the allocating phase.
+class PartialArrayStateManager : public CHeapObj<mtGC> {
+  friend class PartialArrayStateAllocator;
+
+  // Use an arena for each allocator, for thread-safe concurrent allocation by
+  // different allocators.
+  Arena* _arenas;
+
+  // Limit on the number of allocators this manager supports.
+  uint _max_allocators;
+
+  // CounterValues is an integral type large enough to encode a pair of
+  // allocator counters as a single unit for atomic manipulation.
+  using CounterValues = LP64_ONLY(uint64_t) NOT_LP64(uint32_t);
+  using Counter = LP64_ONLY(uint32_t) NOT_LP64(uint16_t);
+
+  union CounterState;
+
+  // A pair of Counters, packaged as an atomic unit.  One counter is the
+  // number of constructed allocators and one is the number of destructed
+  // allocators. The counters are atomic to permit concurrent construction,
+  // and to permit concurrent destruction.  They are an atomic unit to detect
+  // and reject mixing the two phases, without concern for questions of
```

Comment:
It's nice that this library can detect and reject misuse (such as mixing two phases), but I’m not sure why so much effort was spent preventing this. None of the existing users of the library are expected to mix phases in the near future. Could we instead document that mixing two phases is not permitted, and if someone chooses to do so, they do so at their own risk?

Reply:
So far, only 2 of the nearly a dozen(?) potential clients are using this. I'm
not sure that none of them are going to have workers that do some of their
setup after being started. Hence the desire to support concurrency. And if
that, then I feel better about it if there's some usage validation. But maybe
it would be better to just throw a lock at the problem. And if it turns out
none of the use-cases end up needing that concurrency, then I won't object to
a little bit of code simplification.